### ISSUE TITLE
Can't access projects by NAMESPACE/PROJECT_NAME in API

### ISSUE THREAD

**Issue by rhymoid** [gitlabhq/gitlabhq#8290]
The [API docs for the `/projects/:id` endpoint](http://doc.gitlab.com/ce/api/projects.html#get-single-project) mention:

> Get a specific project, identified by project ID or NAMESPACE/PROJECT_NAME, which is owned by the authenticated user. If using namespaced projects call make sure that the NAMESPACE/PROJECT_NAME is URL-encoded, eg. `/api/v3/projects/diaspora%2Fdiaspora` (where `/` is represented by `%2F`).

This seems to be incorrect;

* `/projects/24` works (where `24` is the numeric ID of `not_my_namespace/some_project`, which is accessible by the API user, but not owned by them).
* `/projects/42` works (where `42` is the numeric ID of `personal_space/some_project`, of which the API user is the owner).
* `/projects/not_my_namespace%2Fsome_project` gives a 404.
* `/projects/personal_space%2Fsome_project` gives a 404.
* The above URLs give a 404 no matter how many characters I percent-encode (so this is *not* a duplicate of #6954).

In conclusion: **it appears you can't identify projects by name**.

---

This bug was identified in [GitLab 7.1.1](https://github.com/gitlabhq/gitlabhq/releases/tag/v7.1.1). I'd expect this feature to be available in that version, since #6954 predates that release.

**Comment by bvamos** [comment]
I am one of those who are waiting for this to be resolved :eyes:

**Comment by jubianchi** [comment]
I'm working on this issue for some days and here is where I'm: I found a regex that could allow us to use nearly every possible syntax for project IDs in URLs:

* `/1337` Numerical project ID
* `/namespace/project` Path with namespace ID (not url-encoded)
* `/namespace%2Fproject` Path with namespace ID (url-encoded)

This regular expression will match every IDs format I could guess:

* `123` Numerical project ID
* `123/456` Numerical namespace and project name
* `123/bar` Numerical namespace and project name as string
* `foo/bar` Namespace and project name as strings
* `foo/456` Namespace as string and numerical project name

This get integrated in the API using `requirements`, for example:

```
resource :projects, requirements: { id: Gitlab::Regex.api_project_id_regex('repository') } do
	#...
end
```

_The `repository` argument is used here to make the regular expression match until a defined word (here it id `repository`) which will remove some ambiguities_

I'm at a point where things start to get acceptable but I have a major issue with this regular expression: some identifiers cannot be made unambiguous. For example if the user issue a `GET /api/v3/projects/5/issues`, the expression does not know if the user is asking for project `5/issues` which is allowed or if he is asking for issues of project 5.

BTW, I have some workaround to make the regular expression handle as expected. I'll propose a PR with the changes.

**Comment by rhymoid** [comment]
I'm only asking for providing what the API docs promise, or at least a method to get there. I don't see any benefit whatsoever in supporting `foo/456` or variations on the theme, while it does, as you notice **introduce ambiguities** into the (promised) API.

The same goes for allowing `namespace/project`. If users are hellbent on not percent-encoding a slash, a solution like #7645 would have been a much better option. I would also reject such a solution, because *it's trivial to do percent-encoding*. The API, as it is, has much worse issues.

I think that it would be much better to have an endpoint that can map a (namespaced) project name to its ID. It's a very simple concept, and therefore, KISS.

**Comment by bvamos** [comment]
+1 to rhymod: *"I'm only asking for providing what the API docs promise, or at least a method to get there."*

**Comment by jubianchi** [comment]
@rhymoid 

> I don't see any benefit whatsoever in supporting foo/456

The point here is that we have no choice about this use-case: gitlab allows namespace or project names to be numerical so if the API ever supports using `namespace/project_name` it will have to handle every possible identifier.

> A much better solution for such a simple problem would be an endpoint that provides a mapping

I'm not really OK with this solution as it would force clients to issue at least **2 HTTP requests** to get a single project. Moreover, If we want to keep an up to date mapping( yes, project and namespaces can be renamed) the logix will be even more complex as it will have to be refreshed.

**Comment by rhymoid** [comment]
@jubianchi 

>  gitlab allows namespace or project names to be numerical so if the API ever supports using `namespace/project_name` it will have to handle every possible identifier.

I'm not asking you to support `/project/namespace/project_name`, nor is anyone else in this issue. It would lead to ambiguous URIs. The API documentation explicitly mentions those URIs are not supported.

In my opening post, I was not asking for any specific solution, even. I was simply stating a fact: the implementation and documentation of the API do not match.

There are really two ways to solve this directly:
1. Make the documentation conform to the current implementation, and state that only `/project/24` works.
2. Make the implementation conform to the current documentation, and support `/project/namespace%2Fproject_name` (*note very well: **the slash is percent-encoded**, exactly as stated in the documentation*).

Your solution does neither of the two. It changes the implementation to do something completely different from the documentation, and in turn, even introduces the problem of ambiguity.

> I'm not really OK with this solution as it would force clients to issue at least 2 HTTP requests to get a single project. 

If request count is the issue, then caching is the solution. Again, keep it simple.

> Moreover, If we want to keep an up to date mapping [...] the logix will be even more complex as it will have to be refreshed.

I don't really understand this. Could you provide an example?

**Comment by jubianchi** [comment]
@rhymoid 

I don't see where the implementation and documentation are not conform :

> Make the documentation conform to the current implementation, and state that only /project/24 works.

Sorry, this is false, `namespace%2Fproject` works too

> Make the implementation conform to the current documentation, and support /project/namespace%2Fproject_name

This is already the case : the doc is clear enough (I think) and the implementation works.

> Your solution does neither of the two

In fact, currently it does not but the goal is to experiment to find a solution without the lowest possible number of WTFs and, if possible, support every syntax.
The title of your issue says that you "Can't access projects by NAMESPACE/PROJECT_NAME in API" which is what I'm trying to fix here.

> If request count is the issue, then caching is the solution. Again, keep it simple.

When dealing with API, the request count is IMHO one of the first problem to solve: you could be rate-limited or whatever, but to keep your client fast and make him not load the backend too much, you have to limit calls.

> then caching is the solution. Again, keep it simple.

[TwoHardThings](http://martinfowler.com/bliki/TwoHardThings.html)

The easiest way of doing is, again IMHO, to query with `namespace/project_name` and if the project has been renamed or dropped, to simply return a 404 or  a redirect if the backend provides enough informations.

> I don't really understand this. Could you provide an example?

Given I issue a HTTP request to the mapping endpoint
And I get the following result:
"""
{
	"foo/bar": 42,
	"baz/quxx" : 1337	
}
"""
And I cache this mapping
Then some time later "foo/bar" is renamed to "bar/foo"

What happens ? I'll have to update my local cache, again, it will require too much HTTP request:

* the first request with the old project name `foo/bar` returning a 404
* the second one to update the mapping
* the third one to get the actual project with its new name `bar/foo`

**Comment by bvamos** [comment]
For me on a GitLab 7.5.3:

    H:\>curl --header "PRIVATE-TOKEN: LVJyLP96Q5pJ2qgJreB6" "https://gitlab/api/v3/projects/90"
    {"id":90,...,"name":"GitHookScripts","name_with_namespace":"Automation / GitHookScripts","path":"githookscripts","path_with_namespace":"automation/githookscripts",...}

With name:

    H:\>curl --header "PRIVATE-TOKEN: LVJyLP96Q5pJ2qgJreB6" "https://gitlab/api/v3/projects/automation%2Fgithookscripts"
    <!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
    <html><head>
    <title>404 Not Found</title>
    ...

**Comment by rhymoid** [comment]
> Sorry, this is false, `namespace%2Fproject` works too

It doesn't work for me (7.1.1), it doesn't work for @mabbage (7.2.1), it doesn't work for @evenh (7.5.1).

If you still believe it works, close this issue.

> The title of your issue says that you "Can't access projects by NAMESPACE/PROJECT_NAME in API" which is what I'm trying to fix here.

Please read the documentation quote in the opening post of this issue. I was not referring to the shape of the URIs, just the qualified names of the projects. Those are two entirely different things.

> TwoHardThings

Nebulous quotes are of no value to me. I do value examples of how things can go wrong.

---

Now, regarding your example: I don't understand why you'd want to do that. I might be blinded by my own use case, though.

In my use case, I have to automate some issue and merge request management that relate to the origin of a repository. Really, I only care about that remote repository and the entities that GitLab attaches to the remote repository. That repository, its issues, its merge requests, and whatever else it may have have a permanent ID. So, if I want to refer to what I care about, I should use the ID. Only the very first time, when I try to figure out what the ID is, I look up the project by name.

If I rename the namespace or project, any URI would break, including the Git URI to the remote repository. This is also why there's a non-specific but clear warning in the settings panel about renaming projects. If only I'd remembered the project ID, rather than insisting on resolving its name, then I could just ask the project for its new URI...

---

I'm not suggesting an endpoint that spits out the name-ID mapping. That would be just as useless as my workaround already is, because I would still have to paginate through 300+ entries.

I have a simple and common question for the GitLab instance, so I should get a simple answer. Something like `/please-redirect-me-to?type=project&name=not_my_namespace/some_project` be much more appropriate for ID resolution. Imagine the mess you'd get when DNS would dump zones to clients rather than answering their questions.

**Comment by jubianchi** [comment]
@rhymoid Ok, I now get your use-case more clearly but I still have a question.

> Really, I only care about that remote repository and the entities that GitLab attaches to the remote repository. That repository, its issues, its merge requests, and whatever else it may have have a permanent ID

> Nebulous quotes are of no value to me. I do value examples of how things can go wrong.

That's why I provided an example just below ;)

> Given I issue a HTTP request to the mapping endpoint
> And I get the following result:
> ```
> {
>     "foo/bar": 42,
>     "baz/quxx" : 1337   
> }
> ```
> And I cache this mapping
> Then some time later "foo/bar" is renamed to "bar/foo"

> What happens ? I'll have to update my local cache, again, it will require too much HTTP request:

> the first request with the old project name foo/bar returning a 404
> the second one to update the mapping
> the third one to get the actual project with its new name bar/foo

The same questions apply if the API does not provide any mapping endpoint and you handle this directly in your app.

<hr/>

If i understand correctly, what you care about is having permanent URIs to feed your processes. So, my question is, why would you bother with relying IDs to `namespace/project` identifiers ? I think your tools would be much simpler with a manual configuration based only on project IDs. The resolution logic will then become useless in your project's code.

My concern is a bit different than yours: I integrated Gitlab with many project (https://gitlab.com/laboard/laboard, https://github.com/jubianchi/labpages, and a hubot based bot which will get open-sourced soon) and I want them to be as KISS as possible but let them provide easy to use features and URLs. 

Easy to use, in the end-user point of view, means that URLs are build with information the user knows about: issues/MRs/Milestones IID, project name and namespace, ... not internal/technical IDs nor URL-encoded things. This is to keep the code simple but also provide a good UX. 

That's why I started to reply on this issue, because I thought the use-cases were similar... My bad, I misunderstood your point :)

**Comment by rhymoid** [comment]
> So, my question is, why would you bother with relying IDs to namespace/project identifiers ? I think your tools would be much simpler with a manual configuration based only on project IDs. The resolution logic will then become useless in your project's code.

The problem I have is that my coworkers and I clone almost one hundred repositories through a tool that isn't aware of GitLab (namely: composer). If we want to automate creation of issues and MRs, I have to figure out which IDs our projects have in our GitLab instance.

Since names can change, I prefer to use IDs, and I would very much prefer that the API treats names as weak references (which they are). There is some infrastructure in HTTP that could be used for this, but since most developers only know HTTP superficially, it's better to make it a separate endpoint.

Furthermore, I think consistency is much, much more important than readability in the UX of an API, and is essential to keep a tool simple and stupid. Translating what the user means into what the system should do is exactly the role of most tools. It generally doesn't matter if you use names or IDs, as long as the API is consistent. The only thing that seems consistent in the GitLab API is the use of IDs (not IIDs or names; why can't I name an issue, or refer to a project by IID?), so unless GitLab's API changes radically, it only makes sense that I use IDs rather than names.

> That's why I started to reply on this issue, because I thought the use-cases were similar... My bad, I misunderstood your point :)

Glad we resolved that then :)

**Comment by jubianchi** [comment]
@rhymoid I also extensively use composer with gitlab and would be glad to discuss the issues you have with this setup (somewhere else than here because this is not the topic of this issue) ;)

**Comment by rhymoid** [comment]
Note: it seems that GitLab 7.7.2 (98ef392) does not have this issue.

**Comment by GromNaN** [comment]
For people that encounter this issue: are you using Apache as reverse proxy ?

In that case, you have to configure your Apache vhost to not decode slashes.
```
AllowEncodedSlashes NoDecode
```

There is an exemple on GitLab Recipes.
https://gitlab.com/gitlab-org/gitlab-recipes/commit/73abde7501b6cd128ce22ea9f82dc19fdc867c80

**Comment by rhymoid** [comment]
@GromNaN That seems plausible. Our current 7.7.2 instance uses nginx as a reverse proxy, but our old instance might have used Apache... checking with DevOps.

**Comment by roshangautam** [comment]
@GromNaN helped me fix this. Its more of a server configuration problem than gitlab. All you need to do is add

AllowEncodedSlashes NoDecode

in your vhost configuration and you are good to go.

For more information check this commit by @GromNaN.

https://gitlab.com/gitlab-org/gitlab-recipes/commit/73abde7501b6cd128ce22ea9f82dc19fdc867c80

**Comment by rhymoid** [comment]
Without a response from @mabbage @evenh @bvamos in 12 days, I'm closing this issue. I assume this is exclusively an Apache configuration issue (even though the API's poor design is certainly contributing to the problem).

**Comment by bvamos** [comment]
Sorry for the late reply. **I confirm that it work after I reconfigured my reverse proxy.**
I am using Nginx, so the configuration setting is different:

    proxy_pass http://unix:/home/git/gitlab/tmp/sockets/gitlab.socket:/$request_uri;

I had to add $request_uri to proxy_pass value as a postfix.

**Comment by lbt** [comment]
I use apache as a reverse proxy so I'm adding a note about the 'nocanon' directive for mod_proxy here.
  http://httpd.apache.org/docs/2.2/mod/mod_proxy.html#proxypass

Basically mod_proxy will convert %2F to %252F and you need :

 ProxyPass / http://gitlab-internal:80/ nocanon

Also see http://serverfault.com/questions/455060/cant-get-mod-proxy-to-correctly-forward-encoded-slash-characters-2f

**Comment by letusfly85** [comment]
@GromNaN 

Do you know GitLab with apache 2.2 work well by using `AllowEncodedSlashes on` instead of `AllowEncodedSlashes NoDecode`?

**Comment by stanhu** [comment]
Yes, it should work with `on` since that is the behavior of nginx.

**Comment by letusfly85** [comment]
@stanhu 

Thank you for replying, ummm however I got unexpected error such like below.

```Go
[GIN] 2015/08/01 - 07:21:50 | 200 |    1.128793ms | 10.*.*.*:2239 |   GET     /
[GIN] 2015/08/01 - 07:21:51 | 200 |     519.515us | 10.*.*.*:2270 |   GET     /api/user
[GIN] 2015/08/01 - 07:21:51 | 200 |     670.084us | 10.*.*.*:2272 |   GET     /api/user/repos
[GIN] 2015/08/01 - 07:22:44 | 404 | 15.230345383s | 10.*.*.*:2803 |   POST    /api/repos/hoge/bundle-training
Error #01: invalid character '<' looking for beginning of value
     Meta: Operation aborted
Error #02: invalid character '<' looking for beginning of value
     Meta: Operation aborted
```

I got this error log by using drone.io debug mode.

**Comment by stanhu** [comment]
I'm not sure where that message comes from. Is that error from some golang project?

**Comment by letusfly85** [comment]
@stanhu 

GitLab version is 7.13.0, I build and provisioned  from source,
and using apache 2.2.14 (which doesn't have `NoDecode` option..)

**Comment by letusfly85** [comment]
@stanhu 

>I'm not sure where that message comes from. Is that error from some golang project?
yes, and GitLab didn't show any log with loglevel=info.

**Comment by letusfly85** [comment]
@stanhu 

it's log is drone.io (Golang project log)

we talk about this issue from below.

https://github.com/drone/drone/pull/1112

**Comment by stanhu** [comment]
It has been a to-do of mine to fix GitLab so that it doesn't require `AllowEncodedSlashes On`. I'll work on that soon. Until then you will need to have this option on or branches with slashes may not work properly.

**Comment by letusfly85** [comment]
@stanhu 

Thank you for replying, is there any milestone your fixing? (Of course we will wait :smile: )

>I'll work on that soon. Until then you will need to have this option on or branches with slashes may not work properly.

Do you say that there are some possibilities current GitLab doesn't work even if we configure with `AllowEncodedSlashes On` ?

**Comment by stanhu** [comment]
No, I believe it should work.

**Comment by stanhu** [comment]
I can't promise anything, but I'll try to look at it this weekend for the 7.14 release.

**Comment by letusfly85** [comment]
@stanhu 

Sorry, I changed from `AllowEncodedSlashes on` to `AllowEncodedSlashes On` and retried.
I got below $GITLAB_HOME/log/production.log


```
Started GET "/gitlab/api/v3/projects/hoge/gitlab-itamae" for 10.200.0.227 at 2015-08-01 10:05:40 +0900
Processing by NamespacesController#show as HTML
  Parameters: {"id"=>"api/v3/projects/hoge/gitlab-itamae"}
  Completed 401 Unauthorized in 25ms (ActiveRecord: 12.1ms)
```

Is this log strange?

**Comment by letusfly85** [comment]
>I can't promise anything, but I'll try to look at it this weekend for the 7.14 release.

Thank you! :smile:  I'm looking forward to it!

**Comment by stanhu** [comment]
@letusfly85 That error looks like you don't have access to that project. There are no encoded slashes in the URL. Are you sure you have permissions set correctly?

**Comment by Bugagazavr** [comment]
@stanhu hi, stan look at this `{"id"=>"api/v3/projects/hoge/gitlab-itamae"}` thats is really strange because id must be a `hoge/gitlab-itamae`, looks like all slashes transforms to `%F2`

**Comment by letusfly85** [comment]
@stanhu @Bugagazavr 

>There are no encoded slashes in the URL. Are you sure you have permissions set correctly?

Yes, I have an owner account for this repository.

And, also drone log is difficult to understand, I tried to some debugging with drone.

**Comment by letusfly85** [comment]
@stanhu @Bugagazavr 

@stanhu 

I tried again, and this time I set gitlab loglevel=debug, and I got below.
I used `hoge/bundle-training` this time.

```ruby
  Started GET "/gitlab/api/v3/projects/hoge/bundle-training" for 10.200.0.157 at 2015-08-01 22:55:29 +0900
  Processing by NamespacesController#show as HTML
  Parameters: {"id"=>"api/v3/projects/hoge/bundle-training"}
  (2.3ms)  BEGIN
  (2.3ms)  ROLLBACK
  Namespace Load (2.3ms)  SELECT  `namespaces`.* FROM `namespaces`  WHERE `namespaces`.`path` = 'api/v3/projects/hoge/bundle-training'  ORDER BY `namespaces`.`created_at` DESC, `namespaces`.`id` DESC LIMIT 1
  Completed 401 Unauthorized in 93ms (ActiveRecord: 20.0ms)
  Started GET "/gitlab/api/v3/projects/hoge/bundle-training" for 10.200.0.227 at 2015-08-01 22:55:31 +0900
  Processing by NamespacesController#show as HTML
  Parameters: {"id"=>"api/v3/projects/hoge/bundle-training"}
  (2.2ms)  BEGIN
  (2.3ms)  ROLLBACK
  Namespace Load (2.4ms)  SELECT  `namespaces`.* FROM `namespaces`  WHERE `namespaces`.`path` = 'api/v3/projects/hoge/bundle-training'  ORDER BY `namespaces`.`created_at` DESC, `namespaces`.`id` DESC LIMIT 1
  Completed 401 Unauthorized in 93ms (ActiveRecord: 19.6ms)
  Started GET "/gitlab/users/sign_in" for 10.200.0.227 at 2015-08-01 22:55:34 +0900
  Processing by SessionsController#new as */*
  ApplicationSetting Load (2.4ms)  SELECT  `application_settings`.* FROM `application_settings`   ORDER BY `application_settings`.`id` DESC LIMIT 1
  (2.3ms)  BEGIN
  (2.3ms)  ROLLBACK
```

@Bugagazavr 

and drone side I got again.  this occurred around 
https://github.com/drone/drone/blob/0.4-database/pkg/remote/builtin/gitlab/gitlab.go#L122:L125
https://github.com/Bugagazavr/go-gitlab-client/blob/master/projects.go#L147:L159

```Go
[GIN] 2015/08/01 - 23:06:03 | 404 |  1.806758264s | *.*.*.*:59981 |   POST    /api/repos/hoge/bundle-training
  Error #01: invalid character '<' looking for beginning of value
       Meta: Operation aborted
       Error #02: invalid character '<' looking for beginning of value
            Meta: Operation aborted
```

and used API is below, `/projects/:id`

https://github.com/Bugagazavr/go-gitlab-client/blob/master/projects.go#L12

**Comment by letusfly85** [comment]
and `/projects/:id` is below,

https://github.com/gitlabhq/gitlabhq/blob/ce5928f4df1ebdac1fbc319fd1dbd15585f5da9e/lib/api/helpers.rb#L43:L56

ummmm, something wrong happen before post request to GitLab?? :sweat_drops: 

I'll see drone and go-gitlab-client source code again.

**Comment by stanhu** [comment]
Can use `curl` to make the same request to make sure your client is doing the right thing?

http://doc.gitlab.com/ce/api/

**Comment by Bugagazavr** [comment]
@letusfly85 there is not correct id param `Parameters: {"id"=>"api/v3/projects/hoge/bundle-training"}`,`. If you got this, something going wrong and gitlab try to response with non-api page, it's looks like your apache server transform all slashes to `%F2` symbol.

**Comment by Bugagazavr** [comment]
> Sorry, I changed from AllowEncodedSlashes on to AllowEncodedSlashes On and retried.
maybe you need `AllowEncodedSlashes NoDecode` http://httpd.apache.org/docs/2.2/mod/core.html#allowencodedslashes

**Comment by letusfly85** [comment]
@stanhu @Bugagazavr 

I tried below,

```bash
curl  --insecure --header "Authorization: Bearer 0f46ec326cc171fe31c438d55447872663b55976c9db86715751e6f8bbd53334"  -XGET "https://my.gitlab.com/gitlab/api/v3/projects/hoge%2Fbundle-training"
```

and I got below

```html
<html><body>You are being <a href="https://my.gitlab.com/gitlab/users/sign_in">redirected</a>.</body></html>
```

and I got this oauth_token `0f46ec326cc171fe31c438d55447872663b55976c9db86715751e6f8bbd53334` from drone server which can success logged in with gitlab.

**Comment by letusfly85** [comment]
It seems that I failed to log in with gitlab...

**Comment by Bugagazavr** [comment]
Please try to use private roken fir user, instead oauth token
1 авг. 2015 г. 18:18 пользователь "Shunsuke Wada" <notifications@github.com>
написал:

**Comment by letusfly85** [comment]
@Bugagazavr  @stanhu 

I tried below

```bash
curl  --insecure --header "PRIVATE-TOKEN: j6tgkyUM9uhBr8-3jgCx"  -XGET "https://my.gitlab.com/gitlab/api/v3/projects/hoge%2Fbundle-training"
```

and I got same error message.

Should I refresh private token?? :worried:

**Comment by stanhu** [comment]
Can you just try the basic URL `"http://example.com/api/v3/projects"`?

**Comment by letusfly85** [comment]
@Bugagazavr @stanhu 

Hi, and I tried with not relative_path and not ssl and not apache, only unicorn GitLab.

```bash
curl  --insecure --header "PRIVATE-TOKEN: j6tgkyUM9uhBr8-3jgCx"  -XGET "http://*.*.*.*/api/v3/projects/hoge%2Fbundle-training"
```

I got....


```json
{"id":293,"description":"","default_branch":"master","tag_list":[],"public":false,"archived":false,"visibility_level":0,"ssh_url_to_repo":"git@*.*.*.*:hoge/bundle-training.git","http_url_to_repo":"http://*.*.*.*/hoge/bundle-training.git","web_url":"http://*.*.*.*/hoge/bundle-training","name":"bundle-training","name_with_namespace":"hoge / bundle-training","path":"bundle-training","path_with_namespace":"hoge/bundle-training","issues_enabled":true,"merge_requests_enabled":true,"wiki_enabled":true,"snippets_enabled":false,"created_at":"2015-07-30T06:22:51.000Z","last_activity_at":"2015-07-31T06:19:02.000Z","creator_id":2,"namespace":{"id":10,"name":"hoge","path":"hoge","owner_id":null,"created_at":"2014-10-28T07:37:27.000Z","updated_at":"2014-10-29T06:11:42.000Z","description":"","avatar":{"url":"/uploads/group/avatar/10/hoge.png"}},"avatar_url":null,"permissions":{"project_access":null,"group_access":{"access_level":50,"notification_level":3}}}
```

**Comment by letusfly85** [comment]
@Bugagazavr @stanhu 

and below are successful

```
curl  --insecure --header "PRIVATE-TOKEN: j6tgkyUM9uhBr8-3jgCx"  -XGET "https://my.gitlab.com/gitlab/api/v3/projects/293"
```

```
curl  --insecure --header "Authorization: Bearer 0f46ec326cc171fe31c438d55447872663b55976c9db86715751e6f8bbd53334"  -XGET "https://my.gitlab.com/gitlab/api/v3/projects/293"
```

**Comment by Bugagazavr** [comment]
and when you try to use `namespace/name` you got 401 error?

**Comment by letusfly85** [comment]
@Bugagazavr 


```bash
curl  --insecure --header "Authorization: Bearer 0f46ec326cc171fe31c438d55447872663b55976c9db86715751e6f8bbd53334"  -XGET "http://*.*.*.*/api/v3/projects/namespace%2Fnme"

{"message":"404 Project Not Found"}
```

```bash
curl  --insecure --header "Authorization: Bearer 0f46ec326cc171fe31c438d55447872663b55976c9db86715751e6f8bbd53334"  -XGET "http://*.*.*.*/api/v3/projects/nasaspace/name"

<html><body>You are being <a href="http://*.*.*.*/users/sign_in">redirected</a>.</body></html>
```

**Comment by letusfly85** [comment]
@Bugagazavr @stanhu 


* It seems that `AllowEncodedSlashes On` not working with SSL or relative_path.

* It seems that even if we use encoded slash with curl, gitlab doesn't return expected json with SSL or relative_path.

and I cannot use `AllowEncodedSlashes NoDecode` in production env....

**Comment by letusfly85** [comment]
@Bugagazavr @stanhu 

Ummm, does this test case use correct arguments??

https://github.com/gitlabhq/gitlabhq/blob/master/spec/requests/api/projects_spec.rb#L340:L344

If I have a time, I will see GitLab source from next week :sweat_drops:

**Comment by letusfly85** [comment]
@Bugagazavr @stanhu 

maybe around here, I will debug.

https://github.com/gitlabhq/gitlabhq/blob/ce5928f4df1ebdac1fbc319fd1dbd15585f5da9e/app/models/project.rb#L230:L238

**Comment by stanhu** [comment]
I figu

**Comment by stanhu** [comment]
I figured out the issue. The API code currently only handles a simple `:id` field in `get` definition and doesn't parse the case where you specify a namespace/project combination. If you fix that, the API also expects the namespace **path** and repository **path** (e.g. `root/test`), which isn't documented properly. I'll send a fix.

**Comment by letusfly85** [comment]
@stanhu @Bugagazavr 

Thank you for document fix.

I tried to put some `logger.info(params[:id])` on `lib/api/helper.rb` `def user_project`.
When I use `relative_path` logger doesn't show anything.

Where GitLab figure out `relative_path`after configuration in `config/application.rbb`?

**Comment by stanhu** [comment]
Just to recap, here is what I understand,

1. Currently for the projects API calls, the `id` field **can** be in the form of namespace/project, but this **only** works if this is delivered with an encoded slash (`%2F`). Note that for both the namespace and project, the path, not the name, MUST be used. For example, if you have a project that displays as "Administrator / test", it's possible that the project path is something else (e.g. namespace =`root`, project path = `test`).
2. The omnibus version of GitLab ships with nginx, and the `proxy_pass` configuration is set to pass URLs in an unprocessed form (e.g. `%2F` is left alone).
3. With Apache, multiple configurations have been suggested: `AllowEncodedSlashes On` and `AllowEncodedSlashes NoDecode`. This has generated confusion because the `On` setting has been used to work around issues where branch names have encoded slashes (e.g. https://gitlab.com/gitlab-org/gitlab-ce/issues/1804). 
4. Users attempting to use the namespace/project mechanism for API calls require the `NoDecode` setting.

I've submitted a number of patches to make GitLab handle encoded slashes in branch names (latest addition [here](https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/1084)) to make GitLab work regardless of the setting of `AllowEncodedSlashes`. However, the projects API still requires `NoDecode` to work. This is a bit brittle, and we should improve this.

I originally thought this would be an easy fix by changing the endpoints to use wildcards (e.g. `*id` vs `:id`), but this introduces other issues. For example, in the the API endpoint `/v3/api/projects/namespace/project/events`, the wildcard match would match the entire ID as `projects/namespace/project/events` when "events" should be left alone.

That means the best way to support the namespace/project mode of access would be to define explicit routes for this. But this is a bit of pain since we would need to duplicate routes in many places just to support this, and I doubt this will fly.

Is there any reason why the numeric `:id` shouldn't be the one and only way to access the API, and support for the namespace/project mechanism should be dropped entirely?

**Comment by jubianchi** [comment]
@stanhu I tried to fix this issue some time ago and ended up with what you describe:

> the wildcard match would match the entire ID as projects/namespace/project/events, when "events" should be left alone.

I think the easiest way to solve all the issues is to introduce new routes to the API, for example:

* `/v3/api/projects/:id/events`
* `/v3/api/projects/:namespace/:project/events`

This seemed to require much work when I tried with my Ruby skills. Also, at the time, I was wondering if this should require a new API version or not.

BTW, if I can help on anything here just let me know ;)

**Comment by wdch-nseeley** [comment]
> Is there any reason why the numeric :id shouldn't be the one and only way to access the API, and support for the namespace/project mechanism should be dropped entirely?

@stanhu Trying to download an archive with curl becomes complicated when the project ID is required. Convincing my programmers to discover the project ID before curling is not going to fly here.

I have manually changed the code on my server so the IDs are all properly globbed (*).  One route  conflicts so I have left it alone.

```
GET/PUT    /v3/api/projects/:id/
```

If I glob that route then only that route will work.
If I leave it alone then the rest seem OK.

With an additional route maybe it could be fixed.

```
GET/PUT    /v3/api/projects/:namespace/:project
```

I submitted a pull request with everything globbed.  Obviously its not 100% yet, so perhaps I should delete it.

https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/1205

At the very least, I feel that we should glob everything and leave that one bugged route broken.

Any thoughts?

**Comment by clns** [comment]
It appears there's another issue related to this when using apache as a reverse proxy. When the requests contain an absolute uri, the `nocanon` param on the `ProxyPass` directive doesn't prevent the double encoding of %2F to %252F, as explained here https://github.com/clns/go-gitlab/commit/d5e2e8d0d0c63eebc7450d4e9a8db59354ce2e5c

**Comment by jvanbaarsen** [comment]
We are currently in the process of closing down the issue tracker on GitHub, to prevent duplication with the GitLab.com issue tracker. Since this is an older issue I'll be closing this for now. If you think this is still an issue I encourage you to open it on the [GitLab.com issue tracker](https://gitlab.com/gitlab-org/gitlab-ce/issues).

**Comment by ppaulis** [comment]
This problem gave me some headaches today. So despite that it's closed, here is what fixed it for me:
```
AllowEncodedSlashes NoDecode
PassengerAllowEncodedSlashes On
```
`AllowEncodedSlashes NoDecode` wasn't enough in my case.

**Comment by rondoe** [comment]
the nocanon keyword fixed that for me on apache reverse proxy. thanks a lot.
